[IFC][Ruby] Pass annotation adjustments to IFC

anttijk opened the pull request:
#### a8caffcf64e22d92bb25e5dfd2fcab4768534753
<pre>
[IFC][Ruby] Pass annotation adjustments to IFC
<a href="https://bugs.webkit.org/show_bug.cgi?id=254017">https://bugs.webkit.org/show_bug.cgi?id=254017</a>

Reviewed by Alan Baradlay.

* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::hasAnnotation const):
(WebCore::Layout::InlineLevelBox::annotationAbove const):
(WebCore::Layout::InlineLevelBox::annotationBelow const):
(WebCore::Layout::InlineLevelBox::annotationUnder const): Deleted.

Rename for consistency.

* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxVerticalAligner.cpp:
(WebCore::Layout::LineBoxVerticalAligner::adjustForAnnotationIfNeeded const):

Allow annotation adjustments for atomic inline boxes (because legacy ruby is wrapped to one).

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::computeInkOverflowForInlineBox):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayLineBuilder.cpp:
(WebCore::Layout::InlineDisplayLineBuilder::collectEnclosingLineGeometry const):
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::updateLayoutBoxDimensions):

Pass the annotation adjustments.

* Source/WebCore/layout/layouttree/LayoutBox.cpp:
(WebCore::Layout::Box::rubyAnnotationsAboveAndBelow const):
(WebCore::Layout::Box::setRubyAnnotationsAboveAndBelow):

Add a rare data field.

* Source/WebCore/layout/layouttree/LayoutBox.h:
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::computeOverAnnotationAdjustment const):
(WebCore::LegacyInlineFlowBox::computeUnderAnnotationAdjustment const):

Factor code from here to RenderRubyRun.

* Source/WebCore/rendering/RenderRubyRun.cpp:
(WebCore::RenderRubyRun::annotationsAboveAndBelow const):

Compute the adjustments.

* Source/WebCore/rendering/RenderRubyRun.h:

Canonical link: <a href="https://commits.webkit.org/261788@main">https://commits.webkit.org/261788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7f5122dcb29fca00de6017bc2a1c1e324ad6cb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1536 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4640 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121360 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116902 "Failed to checkout and rebase branch from PR 11603") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23365 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5815 "Passed tests") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118601 "Failed to checkout and rebase branch from PR 11603") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105949 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1152 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46372 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1191 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/98812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15019 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20333 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53184 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8223 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16868 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
```diff
@@ -83,9 +83,9 @@ class InlineLevelBox {
-    std::optional<InlineLayoutUnit> annotationAbove() const { return hasAnnotation() && m_annotation->type == Annotation::Type::Above ? std::make_optional(m_annotation->size) : std::nullopt; }
-    std::optional<InlineLayoutUnit> annotationUnder() const { return hasAnnotation() && m_annotation->type == Annotation::Type::Under ? std::make_optional(m_annotation->size) : std::nullopt; }
+    bool hasAnnotation() const { return (hasContent() || isAtomicInlineLevelBox()) && m_annotation.has_value(); };
+    std::optional<InlineLayoutUnit> annotationAbove() const { return hasAnnotation() ? std::optional { m_annotation->above } : std::nullopt; }
+    std::optional<InlineLayoutUnit> annotationBelow() const { return hasAnnotation() ? std::optional { m_annotation->below } : std::nullopt; }
```
```suggestion
    std::optional<InlineLayoutUnit> annotationAbove() const { return hasAnnotation() ? { m_annotation->above } : std::nullopt; }
    std::optional<InlineLayoutUnit> annotationBelow() const { return hasAnnotation() ? { m_annotation->below } : std::nullopt; }
```

Does this work? or even without the brackets?

anttijk replied:
no

alanbaradlay reviewed (approved):
v.nice

anttijk committed:
[IFC][Ruby] Pass annotation adjustments to IFC
https://bugs.webkit.org/show_bug.cgi?id=254017

Reviewed by Alan Baradlay.

* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::hasAnnotation const):
(WebCore::Layout::InlineLevelBox::annotationAbove const):
(WebCore::Layout::InlineLevelBox::annotationBelow const):
(WebCore::Layout::InlineLevelBox::annotationUnder const): Deleted.

Rename for consistency.

* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxVerticalAligner.cpp:
(WebCore::Layout::LineBoxVerticalAligner::adjustForAnnotationIfNeeded const):

Allow annotation adjustments for atomic inline boxes (because legacy ruby is wrapped to one).

* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::computeInkOverflowForInlineBox):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayLineBuilder.cpp:
(WebCore::Layout::InlineDisplayLineBuilder::collectEnclosingLineGeometry const):
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::updateLayoutBoxDimensions):

Pass the annotation adjustments.

* Source/WebCore/layout/layouttree/LayoutBox.cpp:
(WebCore::Layout::Box::rubyAnnotationsAboveAndBelow const):
(WebCore::Layout::Box::setRubyAnnotationsAboveAndBelow):

Add a rare data field.

* Source/WebCore/layout/layouttree/LayoutBox.h:
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::computeOverAnnotationAdjustment const):
(WebCore::LegacyInlineFlowBox::computeUnderAnnotationAdjustment const):

Factor code from here to RenderRubyRun.

* Source/WebCore/rendering/RenderRubyRun.cpp:
(WebCore::RenderRubyRun::annotationsAboveAndBelow const):

Compute the adjustments.

* Source/WebCore/rendering/RenderRubyRun.h:

Canonical link: https://commits.webkit.org/261788@main
